### ISSUE TITLE
Proposal Design: QueryBuilder

### ISSUE THREAD

**Issue by kyranet** [dirigeants/klasa#287]
### Describe the proposal

The PR #228 went into stale due to the lack of an objective design and lack of a technical specification for QueryBuilder. This issue will remain opened until the design for QueryBuilder gets fully written, specified **and** merged into the master branch.

I'm opening this issue as QueryBuilder is part of the v1.0.0 RoadMap and therefore, its design should start taking priority so it can be written, tested and merged before discord.js 12 does, that way we can land the long-awaited Klasa 1.0.0 with discord.js 12 release.

The current (and last written) version of QueryBuilder looked like this:

```javascript
// Create a QueryBuilder for PostgreSQL
this.qb = new QueryBuilder(this.client, {
	// Declare the boolean type, with the datatype name as "BOOL"
	// and defaults false and true
	BOOLEAN: { name: 'BOOL', default: [false, true] },
	// Declare the float type, defaults to 0
	FLOAT: { name: 'DOUBLE PRECISION', default: 0 },
	// Declare the integer type, defaults to 0
	INTEGER: { name: 'INTEGER', default: 0 },
	// Declare the text type, defaults to NULL
	TEXT: { name: 'TEXT' },
	// Declare the varchar type, defaults to NULL, the property of
	// size tells QB to use VARCHAR(MAX_LENGTH)
	VARCHAR: { name: 'VARCHAR', size: true }
}, {
	// When using arrays, PG converts INTEGER to INTEGER[]. So we want
	// to append '[]' at the end of the type
	arrayWrap: (type) => `${type}[]`
});
```

However, this design has not been approved by the team. I would also like to open this to everyone who wants to contribute in the design of QueryBuilder (perhaps we can take a look to Sequelize and Knex, and get the best of them for some aspects?).

**Comment by MrJacz** [comment]
Shouldn't the types be the same as SettingResolver?
As in we have `boolean` so it would be `boolean: { name: "BOOL", default: [...] }` not `BOOLEAN: { name: 'BOOL', default: [false, true] }`

**Comment by kyranet** [comment]
#306 landed with the final design of QueryBuilder and the approval of the core members. Marking this issue as solved and waiting for #284's merge.
